tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,290 @@
+            }
+
+            return Instance;
+        }
+
+        public async Task StartPollingAsync()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            if (_isPollingStarted)
+            {
+                Log.Warning("Remote Configuration management polling is already started.");
+                return;
+            }
+
+            _isPollingStarted = true;
+            LifetimeManager.Instance.AddShutdownTask(OnShutdown);
+
+            SemaphoreSlim.Release();
+
+            while (!_cancellationSource.IsCancellationRequested)
+            {
+                var isRcmEnabled = !string.IsNullOrEmpty(_discoveryService.ConfigurationEndpoint);
+                var isProductRegistered = _products.Any();
+
+                if (isRcmEnabled && isProductRegistered)
+                {
+                    await Poll().ConfigureAwait(false);
+                    _lastPollError = null;
+                }
+
+                try
+                {
+                    await Task.Delay(_pollInterval, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // We are shutting down, so don't do anything about it
+                }
+            }
+        }
+
+        public void RegisterProduct(Product product)
+        {
+            _products.TryAdd(product.Name, product);
+        }
+
+        public void UnregisterProduct(string productName)
+        {
+            _products.TryRemove(productName, out _);
+        }
+
+        private async Task Poll()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var request = BuildRequest();
+                var response = _remoteConfigurationApi.GetConfigs(request).Result;
+
+                if (response?.Targets?.Signed != null)
+                {
+                    ProcessResponse(response);
+                    _targetsVersion = response.Targets.Signed.Version;
+                }
+            }
+            catch (Exception e)
+            {
+                _lastPollError = e.Message;
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+        }
+
+        private GetRcmRequest BuildRequest()
+        {
+            var appliedConfigurations = _products.Values.SelectMany(pair => pair.AppliedConfigurations.Values).ToList();
+            var cachedTargetFiles = appliedConfigurations.Select(cache => new RcmCachedTargetFile(cache.Path.Path, cache.Length, cache.Hashes.Select(kp => new RcmCachedTargetFileHash(kp.Key, kp.Value)).ToList())).ToList();
+
+            return new GetRcmRequest(BuildRequestClient(), cachedTargetFiles);
+
+            RcmClient BuildRequestClient()
+            {
+                return new RcmClient(_id, _products.Keys.ToList(), BuildRcmClientTracer(), BuildRequestClientState());
```

Comment:
Could we pass the ICollection instead ? IIRC the Json serialization would work the same and that would avoid the extra list allocation

Reply:
Right.
Done at 162c649b444f316c8ef52fd25cbfc0b42b73d939